src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,582 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+      into.start_sequence().encode_list(m_ip_addr_ranges.value()).end_cons();
+   } else {
+      into.encode_null();
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::decode_from(Botan::BER_Decoder& from) {
+   ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   if(next_tag == ASN1_Type::Null) {
+      from.decode_null();
+      m_ip_addr_ranges = std::nullopt;
+   } else if(next_tag == ASN1_Type::Sequence) {
+      std::vector<IPAddressOrRange<V>> ip_ranges;
+      from.decode_list(ip_ranges);
+      m_ip_addr_ranges = ip_ranges;
+   } else {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressChoice {}", static_cast<uint32_t>(next_tag)));
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressChoice<V>::sort_and_merge() {
+   if(!m_ip_addr_ranges.has_value()) {
+      return;
+   }
+   std::vector<IPAddressOrRange<V>> current = m_ip_addr_ranges.value();
+   if(current.empty()) {
+      return;
+   }
+   bool changed = true;
+   while(changed) {
+      changed = false;
+      // sort by the minimum address
+      std::sort(current.begin(), current.end(), [](IPAddressOrRange<V>& a, IPAddressOrRange<V>& b) {
+         return a.min() < b.min();
+      });
+      for(size_t i = 0; i < current.size() - 1; i++) {
+         IPAddressOrRange<V> a = current[i];
+         IPAddressOrRange<V> b = current[i + 1];
+         // only ++ is implemented for IPAddress<V>, instead of +
+         if(b.min() <= a.max() || b.min() == (++a.max())) {
+            // erase old a and b
+            current.erase(current.begin() + i, current.begin() + i + 2);
+            current.insert(current.begin() + i, IPAddressOrRange<V>(a.min(), std::max(a.max(), b.max())));
+            // look again at the new element we just created, might need to be merged with others still
+            i -= 1;
+            changed = true;
+         }
+      }
+   }
+   m_ip_addr_ranges = current;
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressOrRange<V>::encode_into(Botan::DER_Encoder& into) const {
+   const uint8_t version_octets = static_cast<uint8_t>(V);
+
+   std::array<uint8_t, version_octets> min = m_min.value();
+   std::array<uint8_t, version_octets> max = m_max.value();
+
+   uint8_t zeros = 0;
+   uint8_t ones = 0;
+
+   bool zeros_done = false;
+   bool ones_done = false;
+
+   // count contiguos 0s/1s from the right of the min/max addresses
+   for(size_t i = static_cast<size_t>(version_octets); i > 0; i--) {
+      if(!zeros_done) {
+         uint8_t local_zeros = static_cast<uint8_t>(std::countr_zero(min[i - 1]));
+         zeros += local_zeros;
+         zeros_done = (local_zeros != 8);
+      }
+
+      if(!ones_done) {
+         uint8_t local_ones = static_cast<uint8_t>(std::countr_one(max[i - 1]));
+         ones += local_ones;
+         ones_done = (local_ones != 8);
+      }
+
+      if(zeros_done && ones_done) {
+         break;
+      }
+   }
+
```

Comment:
Here again I feel some commentary outlining the general procedure would be helpful